pigosat.go

```diff
@@ -233,6 +234,172 @@ func (p *Pigosat) AddClauses(clauses Formula) {
+// Solve() returns. Valid means they remain 'assumed' until a call to
+// AddClauses(), Assume(), or another Solve(), following the first Solve().
+// They need to stay valid for FailedAssumptions() to return correct values.
+//
+// Example:
+//
+//   p.Assume(1)                // assume unit clause Clause{1, 0}
+//   p.Assume(-2)               // additionally assume unit clause Clause{-2, 0}
+//   res, solution := p.Solve() // assumes 1 and -2 to hold
+//
+//   if res == Unsatisfiable {
+//       if p.FailedAssumption(1) {
+//           // unit clause Clause{1, 0} was necessary to derive Unsatisfiable
+//       }
+//       if p.FailedAssumption(-2) {
+//           // unit clause Clause{-2, 0} was necessary to derive Unsatisfiable
+//       }
+//       // at least one but also both could be necessary
+//
+//       p.Assume(17)  // Previous assumptions are removed. Now assume unit
+//                     // clause Clause{17, 0} for the next call to Solve().
+//
+//   } else if res == Satisfiable {
+//       // The assumptions 1 and -2 hold, so the follow assertion succeeds.
+//       if !(solution[1] && !solution[2]) {
+//           panic("assertion failed")
+//       }
+//
+//       // Assumptions valid, as always, until calls to AddClauses(), etc.
+//       // Further, when entering Solve() the solver knows that the previous
+//       // call returned SAT and it can safely reset the previous assumptions.
+//   } else  {
+//       // res == Unknown
+//
+//       // Assumptions valid, but assignment invalid except for top level
+//       // assigned literals which necessarily need to have this value if the
+//       // formula is SAT
+//
+//       // As above the solver knows that the previous call returned Unknown
+//       // and will reset assumptions before doing anything else.
+//   }
+func (p *Pigosat) Assume(lit Literal) {
+	defer p.ready(false)()
+	C.picosat_assume(p.p, C.int(lit))
+}
+
+// Returns non zero if the literal is a failed assumption, which is defined as
+// an assumption used to derive unsatisfiability. This is as accurate as
+// generating core literals, but still of course is an overapproximation of the
+// set of assumptions really necessary. The technique does not need clausal core
+// generation nor tracing to be enabled and thus can be much more effective. The
+// function can only be called as long the current assumptions are valid. See
+// Assume() for more details.
+func (p *Pigosat) FailedAssumption(lit Literal) bool {
+	defer p.ready(true)()
+	// picoast_failed_assumption SIGABRTs if the following conditional is true
+	if p.Res() != Unsatisfiable || lit == 0 {
+		return false
+	}
+	return C.picosat_failed_assumption(p.p, C.int(lit)) != 0
+}
+
+// Returns a list of failed assumption in the last call to Solve(). It only
+// makes sense if the last call to Solve() returned Unsatisfiable.
+func (p *Pigosat) FailedAssumptions() []Literal {
+	defer p.ready(true)()
```

Comment:
(Moving discussion of your [comment](https://github.com/wkschwartz/pigosat/commit/185788d2cb37f458bf829d238047bb341c471816#commitcomment-19169711) to here to keep the whole PR together.)

The reason why we should lock before checking the last result is to prevent the race condition in which one thread calls `FailedAssumption` while another thread kicks off a new `Solve`. The reason calling `Res` from inside `FailedAssumption` doesn't deadlock is that they both acquire a read lock and neither acquires a write lock. There is not limit on the number of threads holding a read lock. 

The only advantage I can see to avoiding calling a public (i.e., read-lock-acquiring) method such as `Res` here is speed, or avoiding a situation where in the future we want to switch `Res` to acquire a write lock (not sure why we would want to do the latter). In that case, I think it's fine to create an internal method `res` and have all the public methods, including `Res` and `FailedAssumption` call `res`. That change would be cheap, at least, because it's just a couple lines of code.